File: DP/coin_change.cpp
#include<iostream>

using namespace std;

int min(int a, int b);

int main() {
	int C,n;
	cout << "Enter the total value to be changed: ";
	cin >> C;
	cout << "Enter the number of denominatons: ";
	cin >> n;
	int w[n+1];
	cout << "Enter the values of denominations: " << endl;
	for(int i=1;i<=n;i++) {
		cin >> w[i];
	}
	int memo[C+1];
	memset(memo,0,C+1);	
	for(int i=1;i<C+1;i++) {
		int min_no=C+1;
		bool flag=false;
		for(int j=1;j<=n;j++) {
			if(w[j]<=i) {
				flag=true;
				min_no = min(min_no,1+memo[i-w[j]]);

			}
		}
		if(flag) {
			memo[i]=min_no;
		}
	}
	cout << "Memo:" <<endl;
	for(int j=0;j<C+1;j++) {
		cout << memo[j] <<"\t";
	} 
	cout << "\nMin: " << memo[C];
	return 0;

}



int min(int a, int b) {
	return a>b?b:a;
}

File: src/java_codes/dp/CoinChange.java
package java_codes.dp;

import java.util.Arrays;
import java.util.Scanner;

public class CoinChange {

  public static void main(String[] args) {
    Scanner scanner = new Scanner(System.in);
    System.out.print("Enter the total value to be changed: ");
    int C = scanner.nextInt();
    System.out.print("Enter the number of denominations: ");
    int n = scanner.nextInt();
    int[] w = new int[n];
    System.out.println("Enter the values of denominations: ");
    for (int i = 0; i < n; i++) {
      w[i] = scanner.nextInt();
    }

    int[] memo = new int[C + 1];
    Arrays.fill(memo, C + 1); // Initialize with a value larger than any possible number of coins
    memo[0] = 0;

    for (int i = 1; i <= C; i++) {
      for (int j = 0; j < n; j++) {
        if (w[j] <= i) {
          memo[i] = Math.min(memo[i], 1 + memo[i - w[j]]);
        }
      }
    }

    System.out.println("Memo:");
    for (int j = 0; j <= C; j++) {
      System.out.print(memo[j] + "\t");
    }
    System.out.println(
        "\nMin: "
            + (memo[C] > C
                ? -1
                : memo[C])); // If memo[C] is still the large value, it's not possible
    scanner.close();
  }
}
